### PR TITLE
Check against initial values instead of the implicit flag when serializing

### PR DESCRIPTION
**Pull request by darinadler**
#### 2e779d5aabd5f1b5ca6c02e4c4e18230e65e0f50
<pre>
Check against initial values instead of the implicit flag when serializing
<a href="https://bugs.webkit.org/show_bug.cgi?id=250733">https://bugs.webkit.org/show_bug.cgi?id=250733</a>
rdar://104353481

Reviewed by Tim Nguyen and Oriol Brufau.

* LayoutTests/css3/flexbox/css-properties-expected.txt: Updated.
* LayoutTests/css3/flexbox/css-properties.html: Updated to expect proper serialization of the
flex-flow shorthand, omitting unnecessary longhand values as per CSS serializing rules.

* LayoutTests/fast/css/getPropertyValue-columns-expected.txt: Updated.
* LayoutTests/fast/css/getPropertyValue-columns.html: Updated to expect &quot;auto&quot; for the columns
shorthand, omitting the second &quot;auto&quot; longhand value as per CSS serializing rules.

* LayoutTests/fast/css/parse-border-image-repeat-null-crash-expected.txt: Updated to expect
unnecessary &quot;stretch&quot; longhand value to be omitted from border-image shorthand serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt:
Updated failing test lines for changes to the flex-flow shorthand serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/parsing/flex-flow-valid.html:
Updated to expect serialization to omit unnceessary longhand values as per CSS serializing rules.
Specifically, expect &quot;column&quot; instead of &quot;column nowrap&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/list-style-valid-expected.txt:
Updated since more tests pass.

* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html:
Updated to expect serialization to omit unnceessary longhand values as per CSS serializing rules.
Specifically, expect &quot;hidden&quot; instead of &quot;currentcolor hidden medium&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/parsing/outline-valid-mandatory.html:
Updated to allow outline shorthand serialization to serialize outline-color when the value
is entirely default. It would be nice if the specification could be explicit about which longhand
should be serialized in cases like this, so we didn&apos;t have to have so many possible correct
alternatives, since differences can lead to interoperability problems.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt:
Updated since more tests pass.

* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-parsing-valid.html:
Updated to expect serialization to omit unnceessary longhand values as per CSS serializing rules.
Specifically, expect &quot;none reverse&quot; instead of &quot;auto none reverse&quot;.

* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt:
Updated failing test lines for changes to the flex-flow shorthand serialization.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::isAuto): Added for use by valueForOffsetShorthand.
(WebCore::valueForOffsetShorthand): Add logic so we leave out items to follow CSS serializing
rules. Soon we should change our strategy for this for computed style so we can share the
shorthand logic with specified style -- I have figured out an approach and will do it soon -- but
this is needed now to avoid regressions in test results due to improvements in specified style
serialization of the offset shorthand.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const): Moved flex, perspective-origin, and
transform-origin out of the list of shorthands use getShorthandValue, because their syntax does
not allow omitting just any longhand from the serialization, which is what that function does.
Use get2Values for place-content, place-items, and place-self instead of the old
getAlignmentShorthandValue function, matching our current parsing rules for all three. Stop using
getShorthandValue for the container shorthand, since container-type can be omitted, but
container-name cannot. Added serializing code for perspective-origin and transform-origin that
does not omit longhands.
(WebCore::StyleProperties::fontValue const): Added a check for calculated values for font-stretch
since we should not serialize the font shorthand if font-stretch is calculated. Spotted this
error by code inspection while thinking about correct handling of calculated values elsewhere.
(WebCore::StyleProperties::offsetValue const): Rewrote this to check values instead of checking
the implicit flag when deciding what to omit.
(WebCore::StyleProperties::getGridShorthandValue const): Deleted. This was unused.
(WebCore::StyleProperties::getShorthandValue const): Removed the separator argument, since there
is no caller that wants to omit longhands set to the initial value that also wants to use a
separator other than the space character. Use isInitialValueForLonghand instead of
isPropertyImplicit to decide which longhands to omit, and also serialize the first longhand if
all of them have initial values.
(WebCore::StyleProperties::getAlignmentShorthandValue const): Deleted. This was not correct
for any of the cases it was used for.
(WebCore::StyleProperties::borderImagePropertyValue const):  Use isInitialValueForLonghand instead
of isPropertyImplicit to decide which longhands to omit,

* Source/WebCore/css/StyleProperties.h: Removed getGridShorthandValue and
getAlignmentShorthandValue. Removed the separator argument to getShorthandValue.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeColumns): Removed now-unneeded workaround for shorthand
serialization; if both longhands are set to their defaults, we will still serialize as &quot;auto&quot;.
(WebCore::initialValueForLonghand): Added initial values for lots more longhand properties.
Many of these are probably not needed. At one point I thought I needed more, and it seems OK to
have this list be more comprehensive than it needs to be.
(WebCore::isInitialValueForLonghand): Added logic for offset-rotate, which has an initial value
of &quot;auto&quot;, but the parser turns that into a CSSOffsetRotateValue.
(WebCore::initialValueTextForLonghand): Added support for additional initial values that can
now be returned by initialValueForLonghand.

Canonical link: <a href="https://commits.webkit.org/259185@main">https://commits.webkit.org/259185@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d165caabc532b444e393b1831313970406b15efd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104271 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113482 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173771 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108202 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4272 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96495 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112532 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11117 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38777 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92977 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25795 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80441 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6720 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27152 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6854 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3706 "Found 2 new test failures: fast/images/avif-as-image.html, fast/images/avif-image-document.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12869 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46710 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6327 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8640 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html
Comment (Loirooriol) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html:18]: It's not clear how to serialize this case: https://github.com/w3c/csswg-drafts/issues/8155
So I would use an array and keep allowing the old expectation.

File: Source/WebCore/css/StyleProperties.cpp
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp:336]: The grammar is `<'container-name'> [ / <'container-type'> ]?`, so this should omit `container-type` if set to the initial value. That's the case without this patch:

```js
var {style} = document.createElement("div");
style.cssText = "container: foo";
style.container; // "foo"
```

File: Source/WebCore/css/StyleProperties.cpp
Comment (nt1m) [inline Source/WebCore/css/StyleProperties.cpp:359]: What about transform-origin-z ? It should be omitted if equal to 0, but specified otherwise

File: Source/WebCore/css/StyleProperties.cpp
Comment (nt1m) [inline Source/WebCore/css/StyleProperties.cpp:509]: nit: small personal preference, but I find it easier to follow logic when variables are defined just before their first usage.

**Review by nt1m — APPROVED** [review]
I wish we could auto-generate the initial value stuff from CSSPropertyParser.cpp

Looks good to me otherwise

File: Source/WebCore/css/StyleProperties.cpp
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:359]: Yes, fixing this right now.

File: Source/WebCore/css/StyleProperties.cpp
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:509]: I tried various organizations of this function, and this was the one I liked best overall.

**Comment by darinadler** [comment]
> I wish we could auto-generate the initial value stuff from CSSPropertyParser.cpp

We wil! That’s what the comment at the top of the function says.

File: Source/WebCore/css/StyleProperties.cpp
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:509]: My next patch rewrites this function anyway.

File: Source/WebCore/css/StyleProperties.cpp
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:336]: That’s what the line of code above does. If "type is normal, return just the name".

File: LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html
Comment (darinadler) [reply LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html:18]: OK.

The open issue seems to be discussing whether it should be "medium", "hidden", or "current color", not whether it’s OK to return "medium hidden currentcolor". I hope at some point **someone** changes this to not allow "medium hidden currentcolor".

I think we should resolve the many issues like this one. Some of the parsing tests have a comically long list of possible serializations. And others are really strict, only allowing exactly one. I can’t really spot the pattern.

File: Source/WebCore/css/StyleProperties.cpp
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:336]: By the way, while working on this I noticed many shorthands where we do not yet implement the shortest rule. I implemented many of those, it made the patch too large, so I cut down my "shortest" logic to only the amount needed to avoid a regression for now.

Comment (Loirooriol) [reply Source/WebCore/css/StyleProperties.cpp:336]: Ah, sure

File: LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html
Comment (darinadler) [reply LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html:18]: I studied this more, and I realized that the issue you filed, https://github.com/w3c/csswg-drafts/issues/8155, is about the ambiguity when there are multiple minimal representations, like if `medium none currentcolor` is specified it could serialize as `currentcolor`, `none`, or `medium`. This case is not like that, since `hidden` is not the initial value for `column-rule-style`. So there is no other single-word representation. Besides `hidden` we could use `medium hidden` or `hidden currentcolor` or `medium hidden currentcolor`.

Do you still think I should use an array and keep the old expectation? If so, why? Maybe we should not have web platform tests require minimal representations. Some tests do require that, presumably cases where some other engine was already correctly doing a minimal representation, but not WebKit yet. If there are some cases where a non-minimal representation is needed for web compatibility, then we would definitely not want to require the minimal one. But I really don’t know how to decide when to be strict and when to be lenient in WPT.

Comment (darinadler) [reply LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html:18]: @Loirooriol I am going to wait on your answer before merging this in to WebKit or WPT; for now I am not making any change though.

Comment (Loirooriol) [reply LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-valid.html:18]: Ah, I just associated `hidden` with `none` and thought it was the initial value.
But I guess the change is correct according to the shortest serialization principle.

**Review by Loirooriol — APPROVED** [review]

**Comment by nt1m** [comment]
iOS needs a rebaseline

**Comment by nt1m** [comment]
> Check against initial values instead of the implicit flag when serializing
> https://bugs.webkit.org/show_bug.cgi?id=250733
> rdar://104353481 (Check against initial values instead of the implicit flag when serializing (250733))

Btw, you only need the radar link, the title is redundant here :)

**Comment by Loirooriol** [comment]
You should export the WPT changes

**Comment by darinadler** [comment]
Oh, good point, I forgot! Will do that soon.

**Pull request merged**
